bpo-41336: Fix the error handling in zoneinfo_new_instance()

ZackerySpytz opened the pull request:
Do not call PyObject_CallMethod() with a live exception (like
KeyboardInterrupt).

<!-- issue-number: [bpo-41336](https://bugs.python.org/issue41336) -->
https://bugs.python.org/issue41336
<!-- /issue-number -->

serhiy-storchaka commented on Modules/_zoneinfo.c:
```diff
@@ -224,8 +224,14 @@ zoneinfo_new_instance(PyTypeObject *type, PyObject *key)
-        Py_DECREF(tmp);
+        _PyErr_ChainExceptions(exc, val, tb);
+        Py_XDECREF(tmp);
+        if (tmp == NULL) {
```
It is an undefined behavior to compare a pointer to the freed memory with NULL. Do it before `Py_XDECREF(tmp)`.

ZackerySpytz committed:
bpo-41336: Fix the error handling in zoneinfo_new_instance()

Do not call PyObject_CallMethod() with a live exception (like
KeyboardInterrupt).

serhiy-storchaka reviewed (approved):
:+1:

pganssle commented on Modules/_zoneinfo.c:
```diff
@@ -224,8 +224,14 @@ zoneinfo_new_instance(PyTypeObject *type, PyObject *key)
 cleanup:
     if (file_obj != NULL) {
+        PyObject *exc, *val, *tb;
+        PyErr_Fetch(&exc, &val, &tb);
```
What is the purpose of this?

Is the issue that there might be a live exception and that `file_obj.close()` may itself raise an exception, so we want to save the original one and chain it to the new one?

pganssle commented on Modules/_zoneinfo.c:
```diff
@@ -224,8 +224,14 @@ zoneinfo_new_instance(PyTypeObject *type, PyObject *key)
-        Py_DECREF(tmp);
+        _PyErr_ChainExceptions(exc, val, tb);
+        if (tmp == NULL) {
+            Py_CLEAR(self);
```
Hm... this feels a little weird but I suppose it's acceptable.

At this point we've actually already succeeded in reading and constructing a `ZoneInfo` in `self`; if `tmp` is `NULL` that indicates that we've failed to close a file that we've opened for some reason, but we could still return `self`. In what situation would it matter that we failed to close the file handle? I would expect that failure to close the file handle would mean that it's already closed or something has happened to the file descriptor at the OS level — some sort of conflict around ownership of the resource that we could easily cede to whatever other process has taken ownership of it.

Is the idea here that the Ctrl + C here is manifesting as an error in the method call and will only bubble up the stack if we affirmatively raise it here? If so, might it make sense to clear anything derived from `Exception` and bubble up only errors here derived from `BaseException` but not `Exception`?
